File: typescript/src/verification/smt-verification-result.ts
import type { MarkingState } from './marking-state.js';
import type { PInvariant } from './invariant/p-invariant.js';

/**
 * Verification verdict.
 */
export type Verdict = Proven | Violated | Unknown;

/** Property proven safe. No reachable state violates it. */
export interface Proven {
  readonly type: 'proven';
  readonly method: string;
  readonly inductiveInvariant: string | null;
}

/** Property violated. A counterexample trace is available. */
export interface Violated {
  readonly type: 'violated';
}

/** Could not determine. */
export interface Unknown {
  readonly type: 'unknown';
  readonly reason: string;
}

/**
 * Which route decided a verdict ([VER-003]).
 *
 * The routes do equivalent work by different means, and a consumer reading the
 * result's fields rather than its report needs to know which one answered:
 * `enumeration` and `nu-scg` decide by exploring a finite graph and compute no
 * P-invariants at all.
 *
 * The rule for {@link SmtVerificationResult.invariants} is about the *empty* case
 * only: an empty list from a route other than `smt` means "not computed", not
 * "the net has none". A non-empty list is always real — `unavailable` in
 * particular still carries the invariants the pipeline computed before it found
 * no usable solver, and `structural` carries whatever the proof rested on.
 */
export type VerificationRoute =
  /** The IC3/PDR pipeline: flatten, invariants, encode, solve ([VER-001]). */
  | 'smt'
  /** Bounded state-space enumeration ([VER-017]). */
  | 'enumeration'
  /** The ν name-partition state-class graph ([VER-012], Route B). */
  | 'nu-scg'
  /** A structural proof — Commoner's theorem, or the linear bound of [VER-015]. */
  | 'structural'
  /** No route could run (no solver, an unresolved property place). */
  | 'unavailable';

/**
 * Solver statistics.
 */
export interface SmtStatistics {
  readonly places: number;
  readonly transitions: number;
  readonly invariantsFound: number;
  readonly structuralResult: string;
}

/**
 * Result of SMT-based verification.
 */
export interface SmtVerificationResult {
  readonly verdict: Verdict;
  /**
   * Which route decided this verdict ([VER-003]). Read it before concluding
   * anything from an **empty** {@link invariants}: off the `'smt'` route that
   * means "not computed", never "none exist". A non-empty list is real whatever
   * the route says.
   */
  readonly route: VerificationRoute;
  readonly report: string;
  readonly invariants: readonly PInvariant[];
  readonly discoveredInvariants: readonly string[];
  readonly counterexampleTrace: readonly MarkingState[];
  readonly counterexampleTransitions: readonly string[];
  /**
   * Outcome of the abstract counterexample replay, as a TRI-STATE. `null` means
   * "the replay did not apply"; the two booleans both mean it ran.
   *
   * - `true` — an abstract firing chain from M₀ to a property-violating state
   *   was re-executed TS-side; `counterexampleTrace` is that chain in FIRING
   *   (replay) order and the verdict is `violated`.
   * - `false` — the replay ran without confirming the trace. Either it could not
   *   settle the question (nothing decoded from the Z3 derivation, M₀ absent
   *   from the decoded set, or a node/segment budget hit), in which case the
   *   `violated` verdict rests on Spacer's SAT answer alone; or the search
   *   completed and found NO chain, in which case the verdict was downgraded to
   *   `unknown`. The report distinguishes the two ("UNCONFIRMED" vs "FAILED").
   * - `null` — replay did not apply: non-violated verdict, replay disabled via
   *   `counterexampleReplay(false)`, the coloured ν-encoding / Route B (whose
   *   state shapes are outside the flat replayer's scope), or a structural
   *   proof.
   *
   * A `true` from the enumeration route ([VER-017]) means the same thing it means
   * everywhere else — the trace is an ordered firing sequence that reaches the
   * violation — even though it was read off the state-class graph rather than
   * re-executed: the graph path *is* a firing sequence, so there is nothing to
   * re-confirm. Consumers keying "are these steps ordered" off this field get the
   * right answer without special-casing the route.
   */
  readonly counterexampleConfirmed: boolean | null;
  readonly elapsedMs: number;
  readonly statistics: SmtStatistics;
}

export function isProven(result: SmtVerificationResult): boolean {
  return result.verdict.type === 'proven';
}

export function isViolated(result: SmtVerificationResult): boolean {
  return result.verdict.type === 'violated';
}

File: java/src/main/java/org/libpetri/smt/SmtVerificationResult.java
package org.libpetri.smt;

import org.libpetri.analysis.MarkingState;
import org.libpetri.smt.invariant.PInvariant;

import java.time.Duration;
import java.util.List;

/**
 * Result of SMT-based verification.
 *
 * @param verdict                  proven, violated, or unknown
 * @param route                    which route decided this verdict ([VER-003]). Read it before
 *     concluding anything from an <b>empty</b> {@code invariants}: off {@link Route#SMT} that
 *     means "not computed", never "none exist". A non-empty list is real whatever the route says.
 * @param report                   human-readable analysis report
 * @param invariants               P-invariants found during analysis
 * @param discoveredInvariants     IC3-synthesized inductive invariants (empty if not proven by IC3)
 * @param counterexampleTrace      marking trace to error (empty if proven)
 * @param counterexampleTransitions firing sequence to error (empty if proven)
 * @param counterexampleConfirmed  outcome of the abstract counterexample replay
 *     ({@link org.libpetri.smt.z3.AbstractReplayer}), as a TRI-STATE. The
 *     canonical definition lives on the Rust {@code VerificationResult}; this
 *     restates it:
 *     <ul>
 *       <li>{@code TRUE} — the replay chained an abstract firing run from M0 to a
 *           property-violating marking; {@code counterexampleTrace} is that run,
 *           in firing order.</li>
 *       <li>{@code FALSE} — the replay applied and did NOT confirm the
 *           counterexample. Two shapes, reported the same way:
 *           <ul>
 *             <li>it could not conclude (nothing decodable, M0 not among the
 *                 decoded states, or a budget exhausted) — VIOLATED stands,
 *                 unconfirmed, on Spacer's SAT answer alone;</li>
 *             <li>it refuted the trace outright (no firing chain reaches the
 *                 violation) — the verdict is downgraded to UNKNOWN. A refutation
 *                 is strictly more informative than "did not apply", so it reports
 *                 {@code FALSE}, never {@code null}.</li>
 *           </ul></li>
 *       <li>{@code null} — replay did not apply: a non-violated verdict that no
 *           replay touched, replay disabled via
 *           {@link SmtVerifier#counterexampleReplay(boolean)}, or the coloured
 *           &nu;-encoding / Route B path, whose state shapes are outside the flat
 *           replayer's scope.</li>
 *     </ul>
 * @param elapsed                  wall-clock time for verification
 * @param statistics               solver statistics
 */
public record SmtVerificationResult(
    Verdict verdict,
    Route route,
    String report,
    List<PInvariant> invariants,
    List<String> discoveredInvariants,
    List<MarkingState> counterexampleTrace,
    List<String> counterexampleTransitions,
    Boolean counterexampleConfirmed,
    Duration elapsed,
    SmtStatistics statistics
) {

    /**
     * Which route decided a verdict ([VER-003]).
     *
     * <p>The routes do equivalent work by different means, and a consumer reading the
     * result's fields rather than its report needs to know which one answered:
     * {@link #ENUMERATION} and {@link #NU_SCG} decide by exploring a finite graph and
     * compute no P-invariants at all.
     *
     * <p>The rule for {@link SmtVerificationResult#invariants()} is about the <em>empty</em>
     * case only: an empty list from a route other than {@link #SMT} means "not computed",
     * not "the net has none". A non-empty list is always real — {@link #UNAVAILABLE} in
     * particular still carries the invariants the pipeline computed before it found no
     * usable solver, and {@link #STRUCTURAL} carries whatever the proof rested on.
     */
    public enum Route {
        /** The IC3/PDR pipeline: flatten, invariants, encode, solve ([VER-001]). */
        SMT,
        /** Bounded state-space enumeration ([VER-017]). */
        ENUMERATION,
        /** The &nu; name-partition state-class graph ([VER-012], Route B). */
        NU_SCG,
        /** A structural proof — Commoner's theorem, or the linear bound of [VER-015]. */
        STRUCTURAL,
        /** No route could run (no solver, an unresolved property place). */
        UNAVAILABLE
    }

    /**
     * Verification verdict.
     */
    public sealed interface Verdict {
        /**
         * Property proven safe. No reachable state violates it.
         *
         * @param method             how it was proven ("IC3/PDR", "structural", "P-invariant")
         * @param inductiveInvariant the raw IC3-synthesized inductive invariant formula (may be null)
         */
        record Proven(String method, String inductiveInvariant) implements Verdict {}

        /**
         * Property violated. A counterexample trace is available; whether replay
         * confirmed it is carried by
         * {@link SmtVerificationResult#counterexampleConfirmed()}, not by the
         * verdict.
         */
        record Violated() implements Verdict {}

        /**
         * Could not determine.
         *
         * @param reason explanation (timeout, resource limit, etc.)
         */
        record Unknown(String reason) implements Verdict {}
    }

    /**
     * Solver statistics.
     *
     * @param places             number of places in flattened net
     * @param transitions        number of transitions in flattened net
     * @param invariantsFound    number of P-invariants found
     * @param structuralResult   result of structural pre-check
     */
    public record SmtStatistics(
        int places,
        int transitions,
        int invariantsFound,
        String structuralResult
    ) {}

    /**
     * Returns true if the property was proven safe.
     */
    public boolean isProven() {
        return verdict instanceof Verdict.Proven;
    }

    /**
     * Returns true if a counterexample was found.
     */
    public boolean isViolated() {
        return verdict instanceof Verdict.Violated;
    }
}
